ci/cudf_pandas_scripts/pandas-tests/run.sh

```diff
@@ -11,10 +11,15 @@ rapids-logger "Running Pandas tests using $PANDAS_TESTS_BRANCH branch and rapids
 rapids-logger "PR number: ${RAPIDS_REF_NAME:-"unknown"}"
 
 RAPIDS_PY_CUDA_SUFFIX="$(rapids-wheel-ctk-name-gen ${RAPIDS_CUDA_VERSION})"
-RAPIDS_PY_WHEEL_NAME="pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}" rapids-download-wheels-from-s3 ./local-pylibcudf-dep
-RAPIDS_PY_WHEEL_NAME="cudf_${RAPIDS_PY_CUDA_SUFFIX}" rapids-download-wheels-from-s3 ./local-cudf-dep
-python -m pip install $(ls ./local-pylibcudf-dep/pylibcudf*.whl)
-python -m pip install $(ls ./local-cudf-dep/cudf*.whl)[test,pandas-tests]
+
+# Download the cudf and pylibcudf built in the previous step
+RAPIDS_PY_WHEEL_NAME="cudf_${RAPIDS_PY_CUDA_SUFFIX}" rapids-download-wheels-from-s3 ./dist
+RAPIDS_PY_WHEEL_NAME="pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}" rapids-download-wheels-from-s3 ./dist
+
+# echo to expand wildcard before adding `[extra]` requires for pip
+python -m pip install \
+  "$(echo ./dist/cudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test,pandas-tests]" \
+  "$(echo ./dist/pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test,pandas-tests]"
```

Comment:
It wasn't caused by this PR, but let's get these extras fixed here since they aren't quite right. pylibcudf doesn't have a pandas-tests extra. It does have a test extra, but I wouldn't expect it to be needed in this test run.
```suggestion
  "$(echo ./dist/cudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test,pandas-tests]" \
  "$(echo ./dist/pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)"
```

Reply:
It looks like with this combined approach of installation, even though you're explicitly specifying to install pylibcudf that package doesn't get included in dependency resolution for cudf. The two step process might be necessary, or you can switch to adding `--find-links` to the command.

Reply:
I do see the failure at https://github.com/rapidsai/cudf/actions/runs/10425017056/job/28875933840?pr=16575.

```text
./dist./distLooking in indexes: https://pypi.org/simple, https://pypi.anaconda.org/rapidsai-wheels-nightly/simple, https://pypi.nvidia.com/
Processing ./dist/cudf_cu12-24.10.0a205-cp310-cp310-manylinux_2_28_x86_64.whl
Processing ./dist/pylibcudf_cu12-[24](https://github.com/rapidsai/cudf/actions/runs/10425017056/job/28875933840?pr=16575#step:8:25).10.0a205-cp310-cp310-manylinux_2_28_x86_64.whl
...
ERROR: Could not find a version that satisfies the requirement pylibcudf-cu12==24.10.*,>=0.0.0a0 (from cudf-cu12[test]) (from versions: none)
ERROR: No matching distribution found for pylibcudf-cu12==24.10.*,>=0.0.0a0
```

But did you notice that the CUDA 11.8 version of that same job, using the same script, succeeded?

```text
Processing ./dist/cudf_cu11-24.10.0a205-cp311-cp311-manylinux_2_28_aarch64.whl (from cudf-cu11==24.10.0a205)
Processing ./dist/pylibcudf_cu11-[24](https://github.com/rapidsai/cudf/actions/runs/10425017056/job/28875934052?pr=16575#step:8:25).10.0a205-cp311-cp311-manylinux_2_28_aarch64.whl (from pylibcudf-cu11==24.10.0a205)
...
Successfully installed ... cudf-cu11-24.10.0a205 ... pylibcudf-cu11-24.10.0a205 ...
```

([build link](https://github.com/rapidsai/cudf/actions/runs/10425017056/job/28875934052?pr=16575))

If this method of dependency resolution didn't work, I'd have expected those jobs to both fail in the same way. I'm going to merge in latest `branch-24.10` + apply your suggestion, let's see what happens on a re-run.

Reply:
Good point, that thought flashed in and out of my brain while reviewing your PR and left no trace :joy: let's see what we see.

Reply:
It is still failing, so there is something strange happening here.

Reply:
blegh yeah, and same thing... failing for CUDA 12, passing for CUDA 11. I'm gonna pull down the wheels and double-check their metadata. I want to be sure this is just a `pip` behavior I've misunderstood and not a packaging problem.

Reply:
It crossed my mind too that it might be a packaging problem. I glanced at dependencies.yaml and it looks right, but I could have missed something.

What's more likely is that this is a change in pip or Python behavior. Both versions differ between the two jobs: the CUDA 11 job uses Python 3.11 and pip 24.0, while the CUDA 12 job uses Python 3.12 and pip 23.0.1. The pip version seems like a reasonably likely culprit.

Reply:
ah, thanks!! I hadn't noticed the different `pip` versions!!

That's totally it.

I was able to reproduce the failure locally.

<details><summary>code to do that (click me)</summary>

```shell
docker run \
    --rm \
    --env RAPIDS_REPOSITORY=rapidsai/cudf \
    --env _RAPIDS_SHA=7a12b67bd17056d4c980a3c6e89cc8ec9a6bd7c5_ \
    --env RAPIDS_REF_NAME=pull-request/16575 \
    --env RAPIDS_BUILD_TYPE="pull-request" \
    -it rapidsai/citestwheel:cuda12.5.1-ubuntu22.04-py3.10 \
    bash

RAPIDS_PY_CUDA_SUFFIX="$(rapids-wheel-ctk-name-gen ${RAPIDS_CUDA_VERSION})"

# Download the cudf and pylibcudf built in the previous step
RAPIDS_PY_WHEEL_NAME="cudf_${RAPIDS_PY_CUDA_SUFFIX}" rapids-download-wheels-from-s3 ./dist
RAPIDS_PY_WHEEL_NAME="pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}" rapids-download-wheels-from-s3 ./dist

python -m pip install -v \
  "$(echo ./dist/cudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test]" \
  "$(echo ./dist/pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test]"
```

</details>

Inspected the package metadata with `pkginfo`, didn't see any issues... names, versions, and dependency lists all looked correct.

Upgraded `pip` (to 24.2) and tried again:

```shell
pip install --upgrade pip

python -m pip install -v \
  "$(echo ./dist/cudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test]" \
  "$(echo ./dist/pylibcudf_${RAPIDS_PY_CUDA_SUFFIX}*.whl)[test]"
```

That succeeded 🎉 

This makes me think we should be running `pip install --upgrade pip` every time the images at https://github.com/rapidsai/ci-imgs are rebuilt. Can you think of strong reason we **shouldn't** do that? If not, I'll put up a PR proposing that and see what others say.

Reply:
The main argument against updating pip unconditionally everywhere is that we shouldn't require our users to always have the latest version of pip to be able to install our software. This has come up before in devcontainers, for instance, where the latest version of pip available as a system-installable package is significantly older than the latest version available from PyPI, and the question is raised whether it's really a good idea to require a newer one.

For now, for the purpose of CI I think it's OK to always update so that at least we have consistent behavior. At some point we might want to introduce some form of testing to ensure that our packages are actually installable with different versions of pip, though. We have in the past done things (e.g. the rapids-dask-dependency shenanigans) that ultimately require a fairly new version of pip, which could be an onerous constraint on some users.

Reply:
Thanks, ok yeah that makes sense.

I've opened https://github.com/rapidsai/ci-imgs/pull/171 proposing upgrading `pip` in the wheel CI images.

> *At some point we might want to introduce some form of testing to ensure that our packages are actually installable with different versions of pip, though.*

I agree. This type of thing should be tested directly (similar to https://github.com/rapidsai/build-planning/issues/81) if we want to have high confidence in it.

Reply:
Now that the new CI images are published (https://github.com/rapidsai/ci-imgs/actions/runs/10457672191), I've restarted CI here. Merge latest `branch-24.10` into this (figured I might as well, to get a bit more useful test coverage, if I'm going to re-run so many jobs anyway).

Hopefully with newer `pip` across all the CI images, this newer, stricter pattern for these scripts will work.

Reply:
🎉 looks like that worked!

https://github.com/rapidsai/cudf/actions/runs/10458119570/job/28961862201?pr=16575

Thanks a lot for the help, @vyasr !